File: src/ROSHelper.py
#!/usr/bin/env python
import time
import rospy
from sensor_msgs.msg import JointState
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Header
import numpy as np

class ROSHelper():

    def __init__(self):
        self.command_pose_sub = rospy.Subscriber("/command_pose", PoseStamped, self.load_command_pose)
        self.joint_state_pub = rospy.Publisher("/joint_state", JointState, queue_size = 100)
        self.sent_pose_pub = rospy.Publisher("/sent_pose", PoseStamped, queue_size = 100)
        self.measured_pose_pub = rospy.Publisher("/measured_pose", PoseStamped, queue_size = 100)
        self.command_poses = []
        self.max_queued = 0

    def load_command_pose(self, data):
        self.command_poses.append(data)
        self.max_queued = max(len(self.command_poses), self.max_queued)

    def get_command_pose(self):
        if len(self.command_poses) > 0:
            data = self.command_poses[0]
            del self.command_poses[0]
            return data
        else:
            return None

    def publish_joint_state(self, joints):
        self.joint_state_pub.publish(joints)

    def publish_measured_pose(self, pose):
        self.measured_pose_pub.publish(pose)

    def publish_sent_pose(self, pose):
        self.sent_pose_pub.publish(pose)

    def get_max_queued(self):
        return self.max_queued

File: src/ROSHelper.cpp
#include "ROSHelper.hpp"

ROSHelper::ROSHelper(ros::NodeHandle n, int yumi_port)
{
  std::string arm_name;
  if (yumi_port==6510){
    arm_name = "_left";
  }else if(yumi_port==6513){
    arm_name = "_right";
  }else{
    ROS_INFO("[ROSHelper] Wrong port number %d ", yumi_port);
  }
  command_joints_sub = n.subscribe("/command_joints"  + arm_name, 100, &ROSHelper::load_command_joints, this);
  measured_joints_pub = n.advertise<sensor_msgs::JointState>("measured_joints"  + arm_name, 100);
  measured_pose_pub = n.advertise<geometry_msgs::PoseStamped>("measured_pose"  + arm_name, 100);
  sent_joints_pub = n.advertise<sensor_msgs::JointState>("sent_joints"  + arm_name, 100);
}

ROSHelper::~ROSHelper()
{
}

void ROSHelper::load_command_joints(const sensor_msgs::JointState& data)
{
  command_joints = data;
}

sensor_msgs::JointState ROSHelper::get_command_joints()
{
  return command_joints;
}

void ROSHelper::publish_measured_joints(const sensor_msgs::JointState joints)
{
  measured_joints_pub.publish(joints);
}

void ROSHelper::publish_measured_pose(const geometry_msgs::PoseStamped pose)
{
  measured_pose_pub.publish(pose);
}

void ROSHelper::publish_sent_joints(const sensor_msgs::JointState joints)
{
  sent_joints_pub.publish(joints);
}
